### PR TITLE
feat: implement F-score tracking and analytics (issue #91)

### LINKED ISSUE #91: Implement F-score tracking and analytics
**Priority**: CRITICAL  
**Size**: Medium
**Description**: Create dedicated F-score calculation and tracking system for precision/recall metrics as emphasized in the PRD for manager/executive use cases.

**Context**: The PRD explicitly calls out "Track precision/recall metrics with F-scores over time" as a key manager/executive use case. Per architect guidance, we implement the infrastructure in MVP Phase 4 with placeholder/demo data until the full test execution engine is available post-MVP.

**Updated Technical Approach** (Post Type-Driven Refactoring):
- **Leverage new domain types**: Use domain-specific types for all metrics and identifiers
- **Type-safe configuration**: Integration with centralized constants and validation patterns
- **Enhanced error handling**: Use domain-specific error types with validation
- **Test data integration**: Leverage `src/domain/test_data.rs` for consistent demo data

**Acceptance Criteria**:
- **F-score calculation engine (MVP)**:
  - F-score computation framework (harmonic mean) using validated domain types
  - Configurable beta values for F-beta scores with `BatchSize` validation patterns
  - Statistical confidence intervals with type-safe numeric types
  - **Demo data generation** using centralized test data constants from `src/domain/test_data.rs`
- **Time-series tracking infrastructure**:
  - F-score trends over time structure using domain timestamp types
  - Model version correlation framework with `ProviderName` integration
  - Application-specific F-scores design using `ApplicationId` types
  - Environmental F-score comparison structure with `EnvironmentId` types  
- **Analytics integration**:
  - F-score projections in metrics system using validated domain types
  - Dashboard visualization with **placeholder data** (MVP Phase 4)
  - Integration with centralized constants from `src/providers/constants.rs`

**Technical Implementation**:
- Use `src/domain/test_data.rs` constants for demo F-score data generation
- Leverage domain types for application/environment identification 
- Apply validation patterns from network types for confidence intervals
- Use centralized error handling with `ErrorMessage` types
- Integration with EventCore using domain-specific command patterns

**Dependencies**: 
- Domain types system (✅ **COMPLETED** in type-driven refactoring)
- Test data infrastructure (✅ **COMPLETED** in type-driven refactoring)
- Constants system (✅ **COMPLETED** in type-driven refactoring)

### PR DESCRIPTION
**Pull request by jwilger**
## Summary

Implements comprehensive F-score tracking and analytics system as requested in issue #91. This PR introduces dedicated F-score calculation and tracking infrastructure for precision/recall metrics targeting manager/executive use cases in MVP Phase 4.

### Key Features

- **Type-Safe F-score Domain Types**: Created validated newtypes using `nutype` for `FScore`, `Precision`, `Recall`, `Beta`, and `ConfidenceLevel` with proper validation ranges
- **F-score Calculation Engine**: Implements harmonic mean calculation with support for configurable beta values (F1, F2, F-beta scores)
- **EventCore Integration**: Multi-stream event sourcing with `RecordModelFScore` and `RecordApplicationFScore` commands
- **Time-Series Tracking**: `FScoreDataPoint` infrastructure for tracking F-score trends over time with confidence intervals
- **Demo Data Generation**: Comprehensive demo data generator for MVP visualization with realistic time-series data
- **Comprehensive Test Coverage**: 14 new passing tests covering domain logic, EventCore integration, and edge cases

### Technical Implementation

- **Domain Types**: All F-score types include validation ensuring values are finite and within valid ranges (0.0-1.0)
- **Event Sourcing**: Commands emit `FScoreCalculated` and `ApplicationFScoreCalculated` events to appropriate streams
- **State Management**: `MetricsState` tracks F-score history for both model versions and applications
- **Demo Data**: Includes provider comparison data, application-specific trends, and performance categorization

### Recent Refactoring (per PR review)

Based on comprehensive type-driven development review:

1. **Eliminated Primitive Obsession**:
   - Extracted stream ID prefix constants to remove string literals
   - Created `PerformanceCategory` domain type replacing raw strings in demo data
   - Replaced magic numbers in tests with test_data constants

2. **Reduced Code Duplication**:
   - Introduced `FScoreCommand` trait to share common logic between `RecordModelFScore` and `RecordApplicationFScore`
   - Consolidated F-score calculation logic

3. **Improved Type Safety**:
   - Fixed string `contains()` usage on `ApplicationId` with proper equality checks
   - Enhanced readability and maintainability throughout metrics module

### Testing

All 289 tests pass including:
- F-score calculation accuracy with harmonic mean formula
- EventCore command execution and event emission
- Domain type validation and edge cases
- Demo data generation with realistic patterns

Closes #91

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by jwilger** [comment]
## Type-Driven Design Refactoring Complete ✨

I've completed comprehensive type-driven refactoring as requested. Here's what was accomplished:

### 🎯 High Priority Refactoring
- **Created ColorCode domain type** - Validates hex color codes with regex, replacing raw string literals
- **Created Recommendation enum** - Structured variants with context-specific advice instead of raw strings
- **Replaced ALL magic numbers with constants** - Created quality_thresholds module and updated all code to use named constants
- **Enhanced trend analysis types** - Refactored to use domain types properly, added deprecation warnings to backward compatibility methods

### 🏗️ Improved Module Organization
- **Split metrics.rs for better cohesion**:
  - Extracted core metric types (FScore, Precision, Recall, Beta, ConfidenceLevel) to `core_metrics.rs`
  - Moved FScoreDataPoint to its own `data_point.rs` module
  - Extracted MetricsError to separate `errors.rs` module
  - Clear separation of concerns with focused modules

### 🚀 Type Safety Improvements
- All UI-related values now use domain types (ColorCode, Recommendation)
- All thresholds and magic numbers replaced with named constants
- Improved API design with structured recommendation types
- Added deprecation warnings to guide future refactoring

### ✅ All Tests Passing
- 61 metrics tests passing
- Pre-commit checks all green
- No regressions introduced

The codebase now follows strict type-driven development principles with zero primitive obsession in the metrics domain!

File: src/domain/metrics/trend.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    /// Determine trend direction from two metric values
+    pub fn from_values(
+        current: MetricValue,
+        previous: MetricValue,
+        stability_threshold: StabilityThreshold,
+    ) -> Self {
+        // Check if we can calculate percentage change
+        if let Some(percentage_change) = current.percentage_change_from(previous) {
+            // Use percentage change to determine if change is significant
+            if percentage_change.into_inner().abs() <= stability_threshold.into_inner() {
```
[the remark this answers is not in the document]
Comment (jwilger) [reply src/domain/metrics/trend.rs:32]: The units are actually already consistent - both PercentageChange and StabilityThreshold use decimal fractions (e.g., 0.02 = 2%). I've added a clarifying comment to make this explicit.

-- @claude

File: src/domain/metrics/ui_types.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+    /// Get human-readable recommendation text
+    pub fn as_text(&self) -> &'static str {
+        match self {
+            Self::Outstanding { advice } => advice.as_text(),
+            Self::Critical { remediation } => remediation.as_text(),
+            Self::NeedsImprovement {
+                quality_advice,
+                level_context: _,
+            } => quality_advice.as_text(),
```
[the remark this answers is not in the document]
Comment (jwilger) [reply src/domain/metrics/ui_types.rs:110]: The level_context field is actually used for UI display purposes - it provides additional context beyond the quality advice. I've added a comment to clarify that it's intentionally not used in the plain text representation but is valuable for detailed UI displays.

-- @claude

File: src/domain/metrics/demo_data.rs [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+        // Each application has different baseline performance
+        // For demo purposes, we map specific test applications to performance levels
+        let (base_precision, base_recall) =
+            if application_id.as_ref() == test_data::application_ids::MY_APP {
+                (f_scores::HIGH_PRECISION, f_scores::HIGH_RECALL)
+            } else if application_id.as_ref() == test_data::application_ids::MY_APPLICATION {
+                (f_scores::MEDIUM_PRECISION, f_scores::MEDIUM_RECALL)
+            } else {
+                (f_scores::LOW_PRECISION, f_scores::LOW_RECALL)
+            };
```
[the remark this answers is not in the document]
Comment (jwilger) [reply src/domain/metrics/demo_data.rs:179]: I've refactored the string literal comparisons to use match statements as suggested. This improves maintainability and avoids potential typos.

-- @claude

File: src/domain/events.rs [hunk 4]
```diff
@@ -3,11 +3,11 @@
 //! This module defines all domain events that are stored
 //! in the event store using EventCore.
 
-use chrono::{DateTime, Utc};
 use serde::{Deserialize, Serialize};
```
[the remark this answers is not in the document]
Comment (jwilger) [reply src/domain/events.rs:6]: This is acknowledged - the chrono import removal is part of the comprehensive refactoring to use domain Timestamp types throughout the codebase for better type safety.

-- @claude

File: src/domain/metrics/demo_data.rs [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+                    .clamp(
+                        constants::demo_generation::bounds::APPLICATION_MIN_VALUE,
+                        constants::demo_generation::bounds::MAX_VALUE,
+                    ),
+                )
+                .unwrap();
+
+                let sample_increment = SampleIncrement::base();
+                let sample_count = SampleCount::try_new(
+                    f_scores::MEDIUM_SAMPLE + ((point as u64) * sample_increment.into_inner()),
```
[the remark this answers is not in the document]
Comment (jwilger) [reply src/domain/metrics/demo_data.rs:231]: Good catch\! I have updated the calculation to use saturating_mul to prevent potential overflow for large point values.

-- @claude

File: src/domain/events.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jwilger) [reply src/domain/events.rs:6]: This is acknowledged - the chrono import removal is part of the comprehensive refactoring to use domain Timestamp types throughout the codebase for better type safety.

-- @claude
